### DISCUSSION TITLE
[Chapter 03] 자바 리플렉션 기법

### DISCUSSION THREAD

**Discussion by BvrPark** [cowaine/spring-boot-for-msa#16] [Chapter 03. Spring Application Basics]
P. 135 마지막 문단에서 private 키워드가 선언된 변수들은 클래스 외부에서 접근이 자유롭지 못하기 때문에 테스트 케이스를 작성할 때, 
자바 리플렉션 기법이나 테스트 프레임워크를 사용하여 mocking 작업을 해야 한다. 라고 되어있는데

여기서 자바 리플렉션 기법이 어떤 건지 찾아봤는데 런타임 중에 클래스의 정보를 조작하는 기법? 이라는게 뭔가 명확히 개념이 잡히지 않아서
리플렉션 기법이 어떤 건지 궁금합니다.

**Comment by corock** [comment]
리플렉션(_reflection_) 뜻 자체가 반사, 투영이라는 뜻을 가지고 있습니다.
조금 더 정확히 말하면 자바 가상 머신에는 클래스 로더가 있는데, 이를 통해 읽은 클래스 정보를 활용할 수 있는 기법으로 알고 있어요.
예를 들어, `private` 접근 지정자로 설정된 필드조차도 리플렉션을 이용하면 값을 읽거나 세팅할 수 있습니다.

아래는 리플렉션 기법을 이용한 또 다른 예시로 클래스 로더를 통해 `OrderPrinter` 인스턴스를 만들어 버리는 코드입니다.
생성자를 가져올 때 해당 클래스의 매개변수 목록을 배열로 전달하고, 가져온 생성자로 인스턴스를 만들 수 있습니다.

```java
@Slf4j
@SpringBootApplication
public class CorockApplication {

    public static void main(String[] args) throws ClassNotFoundException, NoSuchMethodException, InvocationTargetException, InstantiationException, IllegalAccessException {
        Class<?> aClass = Class.forName("com.cowaine.corock.chapter03.di.OrderPrinter");
        Constructor<?> constructor = aClass.getConstructor(Formatter.class);
        OrderPrinter orderPrinter = (OrderPrinter) constructor.newInstance(new LocalDateTimeFormatter());
        log.info(orderPrinter.toString());
    }

}
```

그냥 인스턴스 만드는 거 아니야? 라고 생각할 수 있지만, `OrderPrinter` 클래스에서 생성자를 막은 점을 주목해야 합니다.

```java
@Service
@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
@ToString
public class OrderPrinter implements Printer<OrderProduct> {

    private Formatter<LocalDateTime> formatter;

    // (...)

}
```

애플리케이션을 실행하면, 터미널에 찍힌 로그를 통해 정상적으로 인스턴스가 생성된 것을 확인할 수 있습니다.

```bash
2023-08-22 19:25:11.711  INFO 12644 --- [           main] com.cowaine.corock.CorockApplication     : OrderPrinter(formatter=LocalDateTimeFormatter(formatter=Value(YearOfEra,4,19,EXCEEDS_PAD)'-'Value(MonthOfYear,2)'-'Value(DayOfMonth,2)'T'Value(HourOfDay,2)':'Value(MinuteOfHour,2)':'Value(SecondOfMinute,2)))
```

스프링 웹 개발을 한다면 다음과 같은 클래스를 참고하는 것도 도움이 될 것 같습니다.
(`org.springframework.util.ReflectionUtils`, `org.springframework.test.util.ReflectionTestUtils`)

**Comment by JoisFe** [comment]
`@RequestBody`로 json 데이터를 jackson llibrary로 DTO 객체에 마셜링 하는 원리를 알아보시면 좋을 것 같습니다.

**Comment by coalong** [comment]
리플렉션을 직접적으로 사용하기보다 라이브러리에서 많은 예시를 볼 수 있는 것 같습니다.
그와 관련한 예시 ModelMapper 라이브러리 정리 링크를 남깁니다.

https://frost-canvas-b44.notion.site/Mapstruct-7307e37839f34b279ff269a9c2ced991?pvs=4
